[GWT] Fix: Reflection cache generator stopped logging types.

DaanVanYperen opened the pull request:
PR fixes logging in `ReflectionCacheGenerator` which was broken in 90fc8f7
The logging is useful for spotting configuration issues with the reflection cache.

#### Reproduction steps/code
1. Build html target to generate reflection cache and look at the logs. (Any project should do).

Observed: Logging shows empty lines at the beginning (see below).
Expected: List of types included in the reflection cache 
```
[INFO]    Computing all possible rebind results for 'com.badlogic.gwtref.client.IReflectionCache'
[INFO]       Rebinding com.badlogic.gwtref.client.IReflectionCache
[INFO]          Invoking generator com.badlogic.gwtref.gen.ReflectionCacheGenerator
[INFO]             com.badlogic.gwtref.client.IReflectionCache
[INFO]               
[INFO]                 
[INFO]               
[INFO]                 
..
[INFO]            
etc         
```
#### Version of libGDX and/or relevant dependencies
1.10.1-SNAPSHOT

tommyettinger commented:
Hm, that printing code from April 2020 definitely looks wrong, since it ignores its parameter `message`... But I haven't noticed this particular issue in 1.10.0, or any of the versions that should have been affected. I actually get no info logs when I run `html:dist`, just warn and error (if any). Is there some step I need to do to get `[INFO]` logs to appear during compilation? I don't think `Gdx.app.setLogLevel()` could be used here, because this is before Gdx.app has been assigned, right? I'd like to test this out, I'm just not sure if something is wrong with how I would do so.

I don't see any issue with merging; ignoring `message` as 1.10.0 does is a pretty clear bug, and I'm not sure why the original PR was merged as it was, with no discussion. I'd like to know how to get the info logs to show up over here, though!

tommyettinger reviewed (approved):
This is a pretty clear fix for an omission that's lasted over a year. *Stamps with approval.* I think most people probably don't see info logs, otherwise this would have been noticed earlier...

MrStahlfelge reviewed (approved):
This proofed that GWT build gives to much output no one really reads, but some do.

DaanVanYperen commented:
> actually get no info logs when I run `html:dist`, just warn and error (if any).

Sorry, my mistake, should have tested it with a fresh gdx-setup project.

I assume the difference is this `html/build.gradle`
```groovy
gwt {
    ..   
    logLevel = 'INFO' // or TRACE
```

> This proofed that GWT build gives to much output no one really reads, but some do.

It's definitely too verbose for anything but problem solving. Especially with the generation running twice now. Helpful when things break though.

Did play around a bit with a more readable summary @ artemis-odb. Not certain if it adds enough value at this time.

```
[INFO]             Reflected types summary
[INFO]               <No package>.* (types: 7)
[INFO]               com.artemis.* (types: 34)
[INFO]               com.artemis.Aspect.* (types: 1)
[INFO]               com.artemis.ComponentTypeFactory.* (types: 1)
[INFO]               com.artemis.EntitySubscription.* (types: 2)
[INFO]               com.artemis.EntitySubscriptionTest.* (types: 6)
[INFO]               com.artemis.WorldConfigurationBuilder.* (types: 1)
[INFO]               com.artemis.annotations.* (types: 6)
[INFO]               com.artemis.component.* (types: 30)
[INFO]               com.artemis.components.* (types: 1)
[INFO]               com.artemis.injection.* (types: 5)
[INFO]               com.artemis.io.* (types: 20)
[INFO]               com.artemis.link.* (types: 9)
[INFO]               com.artemis.managers.* (types: 13)
[INFO]               com.artemis.prefab.* (types: 6)
[INFO]               com.artemis.system.* (types: 4)
[INFO]               com.artemis.systems.* (types: 8)
[INFO]               com.artemis.utils.* (types: 4)
[INFO]               com.badlogic.gdx.* (types: 123)
[INFO]             287 types reflected for **IReflectionCacheGenerated**

[INFO]             Reflected types summary
[INFO]               <No package>.* (types: 11)
[INFO]               com.badlogic.gdx.* (types: 215)
[INFO]               com.google.gwt.* (types: 25)
[INFO]               java.lang.* (types: 18)
[INFO]               java.util.* (types: 9)
[INFO]               junit.framework.* (types: 4)
[INFO]             287 types reflected for **IReflectionCache2Generated**
```

tommyettinger commented:
Thanks, I didn't know or remember `gwt.logLevel`. Looking at that output, I'm curious what would be putting types in the default <no package> path... Anything in libGDX or Artemis? Or was that your testing code? I think it's probably better to be able to see the actual classes, especially at a high-detail log level.

DaanVanYperen commented:
Looking at that output, I'm curious what would be putting types in the default path

The overview is just grouping packages under a wildcard. Below is the more verbose output for a blank `gdx-setup` project.

> I think it's probably better to be able to see the actual classes, especially at a high-detail log level.

Yeah does that as well. Goal with the at-a-glance was to hint developers when includes are too broad. Relates to Reflection cache preventing GWT dead code elimination (Junkdog/artemis-odb#645).

### Summary GDX-Setup reflection cache

```
            Reflected types summary
              <No package>.* (types: 7)
              com.badlogic.gdx.* (types: 209)
            com.badlogic.gwtref.client.IReflectionCache
```

```
            Reflected types summary
              <No package>.* (types: 13)
              com.badlogic.gdx.* (types: 162)
              com.google.gwt.* (types: 18)
              java.lang.* (types: 18)
              java.util.* (types: 7)
            com.badlogic.gwtref.client.IReflectionCache2
```

### Verbose GDX reflection cache

```
            Reflected types per package
              <No package> (types: 7)
                boolean
                boolean[]
                byte
                byte[]
                byte[][]
                char
                char[]
              com.badlogic.gdx (types: 1)
                Net
              com.badlogic.gdx.Net (types: 6)
                HttpMethods
                HttpRequest
                HttpResponse
                HttpResponseListener
                Protocol
                Protocol[]
              com.badlogic.gdx.graphics (types: 7)
                Color
                Colors
                Texture
                TextureData
                VertexAttribute
                VertexAttribute[]
                VertexAttributes
              com.badlogic.gdx.graphics.Texture (types: 4)
                TextureFilter
                TextureFilter[]
                TextureWrap
                TextureWrap[]
              com.badlogic.gdx.graphics.TextureData (types: 3)
                Factory
                TextureDataType
                TextureDataType[]
              com.badlogic.gdx.graphics.VertexAttributes (types: 1)
                Usage
              com.badlogic.gdx.graphics.g2d (types: 10)
                BitmapFont
                BitmapFontCache
                GlyphLayout
                NinePatch
                Sprite
                SpriteBatch
                SpriteCache
                TextureRegion
                TextureRegion[]
                TextureRegion[][]
              com.badlogic.gdx.graphics.g2d.BitmapFont (types: 3)
                BitmapFontData
                Glyph
                Glyph[][]
              com.badlogic.gdx.graphics.g2d.GlyphLayout (types: 1)
                GlyphRun
              com.badlogic.gdx.graphics.g2d.TextureAtlas (types: 1)
                AtlasRegion
              com.badlogic.gdx.graphics.g3d.model (types: 6)
                Animation
                MeshPart
                Node
                NodeAnimation
                NodeKeyframe
                NodePart
              com.badlogic.gdx.graphics.g3d.model.data (types: 13)
                ModelAnimation
                ModelData
                ModelMaterial
                ModelMesh
                ModelMeshPart
                ModelMeshPart[]
                ModelNode
                ModelNodeAnimation
                ModelNodeKeyframe
                ModelNodePart
                ModelNodePart[]
                ModelNode[]
                ModelTexture
              com.badlogic.gdx.graphics.g3d.model.data.ModelMaterial (types: 2)
                MaterialType
                MaterialType[]
              com.badlogic.gdx.graphics.g3d.particles (types: 6)
                ParallelArray
                ParticleController
                ParticleControllerComponent
                ParticleEffect
                ParticleEffectLoader
                ResourceData
              com.badlogic.gdx.graphics.g3d.particles.ParallelArray (types: 6)
                Channel
                ChannelDescriptor
                ChannelInitializer
                FloatChannel
                IntChannel
                ObjectChannel
              com.badlogic.gdx.graphics.g3d.particles.ParticleEffectLoader (types: 2)
                ParticleEffectLoadParameter
                ParticleEffectSaveParameter
              com.badlogic.gdx.graphics.g3d.particles.ResourceData (types: 3)
                AssetData
                Configurable
                SaveData
              com.badlogic.gdx.graphics.g3d.particles.emitters (types: 2)
                Emitter
                RegularEmitter
              com.badlogic.gdx.graphics.g3d.particles.emitters.RegularEmitter (types: 2)
                EmissionMode
                EmissionMode[]
              com.badlogic.gdx.graphics.g3d.particles.influencers (types: 11)
                ColorInfluencer
                DynamicsInfluencer
                DynamicsModifier
                Influencer
                ModelInfluencer
                ParticleControllerFinalizerInfluencer
                ParticleControllerInfluencer
                RegionInfluencer
                ScaleInfluencer
                SimpleInfluencer
                SpawnInfluencer
              com.badlogic.gdx.graphics.g3d.particles.influencers.ColorInfluencer (types: 2)
                Random
                Single
              com.badlogic.gdx.graphics.g3d.particles.influencers.DynamicsModifier (types: 9)
                Angular
                BrownianAcceleration
                CentripetalAcceleration
                FaceDirection
                PolarAcceleration
                Rotational2D
                Rotational3D
                Strength
                TangentialAcceleration
              com.badlogic.gdx.graphics.g3d.particles.influencers.ModelInfluencer (types: 2)
                Random
                Single
              com.badlogic.gdx.graphics.g3d.particles.influencers.ParticleControllerInfluencer (types: 2)
                Random
                Single
              com.badlogic.gdx.graphics.g3d.particles.influencers.RegionInfluencer (types: 4)
                Animated
                AspectTextureRegion
                Random
                Single
              com.badlogic.gdx.graphics.g3d.particles.renderers (types: 9)
                BillboardControllerRenderData
                BillboardRenderer
                ModelInstanceControllerRenderData
                ModelInstanceRenderer
                ParticleControllerControllerRenderer
                ParticleControllerRenderData
                ParticleControllerRenderer
                PointSpriteControllerRenderData
                PointSpriteRenderer
              com.badlogic.gdx.graphics.g3d.particles.values (types: 15)
                CylinderSpawnShapeValue
                EllipseSpawnShapeValue
                GradientColorValue
                LineSpawnShapeValue
                MeshSpawnShapeValue
                NumericValue
                ParticleValue
                PointSpawnShapeValue
                PrimitiveSpawnShapeValue
                RangedNumericValue
                RectangleSpawnShapeValue
                ScaledNumericValue
                SpawnShapeValue
                UnweightedMeshSpawnShapeValue
                WeightMeshSpawnShapeValue
              com.badlogic.gdx.graphics.g3d.particles.values.MeshSpawnShapeValue (types: 1)
                Triangle
              com.badlogic.gdx.graphics.g3d.particles.values.PrimitiveSpawnShapeValue (types: 2)
                SpawnSide
                SpawnSide[]
              com.badlogic.gdx.maps (types: 2)
                MapObject
                MapObjects
              com.badlogic.gdx.maps.objects (types: 6)
                CircleMapObject
                EllipseMapObject
                PolygonMapObject
                PolylineMapObject
                RectangleMapObject
                TextureMapObject
              com.badlogic.gdx.maps.tiled.objects (types: 1)
                TiledMapTileMapObject
              com.badlogic.gdx.math (types: 40)
                Affine2
                BSpline
                Bezier
                Bresenham2
                CatmullRomSpline
                Circle
                ConvexHull
                CumulativeDistribution
                DelaunayTriangulator
                EarClippingTriangulator
                Ellipse
                FloatCounter
                Frustum
                GeometryUtils
                GridPoint2
                GridPoint3
                Interpolation
                Intersector
                MathUtils
                Matrix3
                Matrix4
                Matrix4[]
                Octree
                Path
                Plane
                Plane[]
                Polygon
                Polyline
                Quaternion
                Quaternion[]
                RandomXS128
                Rectangle
                Shape2D
                Vector
                Vector2
                Vector2[]
                Vector3
                Vector3[]
                Vector[]
                WindowedMean
              com.badlogic.gdx.math.Interpolation (types: 15)
                Bounce
                BounceIn
                BounceOut
                Elastic
                ElasticIn
                ElasticOut
                Exp
                ExpIn
                ExpOut
                Pow
                PowIn
                PowOut
                Swing
                SwingIn
                SwingOut
              com.badlogic.gdx.math.Intersector (types: 2)
                MinimumTranslationVector
                SplitTriangle
              com.badlogic.gdx.math.Octree (types: 2)
                Collider
                RayCastResult
              com.badlogic.gdx.math.Plane (types: 2)
                PlaneSide
                PlaneSide[]
              com.badlogic.gdx.math.collision (types: 4)
                BoundingBox
                Ray
                Segment
                Sphere
              com.badlogic.gdx.scenes.scene2d (types: 5)
                Action
                Action[]
                Actor
                Actor[]
                Event
            Reflected types summary
              <No package>.* (types: 7)
              com.badlogic.gdx.* (types: 209)
            com.badlogic.gwtref.client.IReflectionCache
```

```
            Reflected types per package
              <No package> (types: 13)
                double
                double[]
                float
                float[]
                float[][]
                int
                int[]
                int[][]
                long
                long[]
                short
                short[]
                void
              com.badlogic.gdx.scenes.scene2d (types: 7)
                EventListener
                Group
                InputEvent
                InputListener
                Stage
                Touchable
                Touchable[]
              com.badlogic.gdx.scenes.scene2d.InputEvent (types: 2)
                Type
                Type[]
              com.badlogic.gdx.scenes.scene2d.Stage (types: 1)
                TouchFocus
              com.badlogic.gdx.scenes.scene2d.actions (types: 33)
                Actions
                AddAction
                AddListenerAction
                AfterAction
                AlphaAction
                ColorAction
                CountdownEventAction
                DelayAction
                DelegateAction
                EventAction
                FloatAction
                IntAction
                LayoutAction
                MoveByAction
                MoveToAction
                ParallelAction
                RelativeTemporalAction
                RemoveAction
                RemoveActorAction
                RemoveListenerAction
                RepeatAction
                RotateByAction
                RotateToAction
                RunnableAction
                ScaleByAction
                ScaleToAction
                SequenceAction
                SizeByAction
                SizeToAction
                TemporalAction
                TimeScaleAction
                TouchableAction
                VisibleAction
              com.badlogic.gdx.scenes.scene2d.ui (types: 35)
                Button
                ButtonGroup
                Button[]
                Cell
                CheckBox
                Container
                Dialog
                HorizontalGroup
                Image
                ImageButton
                ImageTextButton
                Label
                List
                ParticleEffectActor
                ProgressBar
                ScrollPane
                SelectBox
                Skin
                Slider
                SplitPane
                Stack
                Table
                TextArea
                TextButton
                TextField
                TextTooltip
                Tooltip
                TooltipManager
                Touchpad
                Tree
                Value
                VerticalGroup
                Widget
                WidgetGroup
                Window
              com.badlogic.gdx.scenes.scene2d.ui.Button (types: 1)
                ButtonStyle
              com.badlogic.gdx.scenes.scene2d.ui.CheckBox (types: 1)
                CheckBoxStyle
              com.badlogic.gdx.scenes.scene2d.ui.ImageButton (types: 1)
                ImageButtonStyle
              com.badlogic.gdx.scenes.scene2d.ui.ImageTextButton (types: 1)
                ImageTextButtonStyle
              com.badlogic.gdx.scenes.scene2d.ui.Label (types: 1)
                LabelStyle
              com.badlogic.gdx.scenes.scene2d.ui.List (types: 1)
                ListStyle
              com.badlogic.gdx.scenes.scene2d.ui.ProgressBar (types: 1)
                ProgressBarStyle
              com.badlogic.gdx.scenes.scene2d.ui.ScrollPane (types: 1)
                ScrollPaneStyle
              com.badlogic.gdx.scenes.scene2d.ui.SelectBox (types: 1)
                SelectBoxStyle
              com.badlogic.gdx.scenes.scene2d.ui.Skin (types: 1)
                TintedDrawable
              com.badlogic.gdx.scenes.scene2d.ui.Slider (types: 1)
                SliderStyle
              com.badlogic.gdx.scenes.scene2d.ui.SplitPane (types: 1)
                SplitPaneStyle
              com.badlogic.gdx.scenes.scene2d.ui.Table (types: 3)
                Debug
                DebugRect
                Debug[]
              com.badlogic.gdx.scenes.scene2d.ui.TextArea (types: 1)
                TextAreaListener
              com.badlogic.gdx.scenes.scene2d.ui.TextButton (types: 1)
                TextButtonStyle
              com.badlogic.gdx.scenes.scene2d.ui.TextField (types: 6)
                DefaultOnscreenKeyboard
                OnscreenKeyboard
                TextFieldClickListener
                TextFieldFilter
                TextFieldListener
                TextFieldStyle
              com.badlogic.gdx.scenes.scene2d.ui.TextField.TextFieldFilter (types: 1)
                DigitsOnlyFilter
              com.badlogic.gdx.scenes.scene2d.ui.TextTooltip (types: 1)
                TextTooltipStyle
              com.badlogic.gdx.scenes.scene2d.ui.Touchpad (types: 1)
                TouchpadStyle
              com.badlogic.gdx.scenes.scene2d.ui.Tree (types: 2)
                Node
                TreeStyle
              com.badlogic.gdx.scenes.scene2d.ui.Value (types: 2)
                Fixed
                Fixed[]
              com.badlogic.gdx.scenes.scene2d.ui.Window (types: 1)
                WindowStyle
              com.badlogic.gdx.scenes.scene2d.utils (types: 21)
                ActorGestureListener
                ArraySelection
                BaseDrawable
                ChangeListener
                ClickListener
                Cullable
                Disableable
                DragAndDrop
                DragListener
                DragScrollListener
                Drawable
                FocusListener
                Layout
                NinePatchDrawable
                ScissorStack
                Selection
                SpriteDrawable
                TextureRegionDrawable
                TiledDrawable
                TransformDrawable
                UIUtils
              com.badlogic.gdx.scenes.scene2d.utils.ChangeListener (types: 1)
                ChangeEvent
              com.badlogic.gdx.scenes.scene2d.utils.DragAndDrop (types: 3)
                Payload
                Source
                Target
              com.badlogic.gdx.scenes.scene2d.utils.FocusListener (types: 1)
                FocusEvent
              com.badlogic.gdx.scenes.scene2d.utils.FocusListener.FocusEvent (types: 2)
                Type
                Type[]
              com.badlogic.gdx.utils (types: 10)
                Array
                ArrayMap
                Disposable
                Json
                JsonReader
                JsonValue
                JsonWriter
                ObjectMap
                OrderedMap
                Queue
              com.badlogic.gdx.utils.Array (types: 2)
                ArrayIterable
                ArrayIterator
              com.badlogic.gdx.utils.ArrayMap (types: 3)
                Entries
                Keys
                Values
              com.badlogic.gdx.utils.Json (types: 3)
                ReadOnlySerializer
                Serializable
                Serializer
              com.badlogic.gdx.utils.JsonValue (types: 4)
                JsonIterator
                PrettyPrintSettings
                ValueType
                ValueType[]
              com.badlogic.gdx.utils.JsonWriter (types: 2)
                OutputType
                OutputType[]
              com.badlogic.gdx.utils.ObjectMap (types: 4)
                Entries
                Entry
                Keys
                Values
              com.badlogic.gdx.utils.Queue (types: 2)
                QueueIterable
                QueueIterator
              com.google.gwt.user.client.rpc.core.java.lang (types: 9)
                Boolean_CustomFieldSerializer
                Byte_CustomFieldSerializer
                Character_CustomFieldSerializer
                Double_CustomFieldSerializer
                Float_CustomFieldSerializer
                Integer_CustomFieldSerializer
                Object_Array_CustomFieldSerializer
                Short_CustomFieldSerializer
                String_CustomFieldSerializer
              com.google.gwt.user.client.rpc.core.java.util (types: 5)
                ArrayList_CustomFieldSerializer
                Collection_CustomFieldSerializerBase
                Collections
                HashMap_CustomFieldSerializer
                Map_CustomFieldSerializerBase
              com.google.gwt.user.client.rpc.core.java.util.Collections (types: 4)
                EmptyList_CustomFieldSerializer
                EmptyMap_CustomFieldSerializer
                EmptySet_CustomFieldSerializer
                SingletonList_CustomFieldSerializer
              java.lang (types: 18)
                Boolean
                Byte
                CharSequence
                CharSequence[]
                Character
                Double
                Enum
                Enum[]
                Float
                Integer
                Object
                Object[]
                Short
                String
                StringBuffer
                StringBuilder
                StringIndexOutOfBoundsException
                String[]
              java.util (types: 7)
                ArrayList
                Collection
                Collections
                HashMap
                List
                ListIterator
                Map
            Reflected types summary
              <No package>.* (types: 13)
              com.badlogic.gdx.* (types: 162)
              com.google.gwt.* (types: 18)
              java.lang.* (types: 18)
              java.util.* (types: 7)
            com.badlogic.gwtref.client.IReflectionCache2
```

tommyettinger commented:
Err, am I reading this right? Was that real output where it used two cache classes for a default gdx-setup app? (Or was that just sample output to show what the IReflectionCache2 would look like?) I'd be surprised if small projects needed two caches, unless there is some benefit from having two smaller caches vs. one larger cache. I didn't think there was such a benefit, since I thought the first cache had to be checked before the second could be used.

DaanVanYperen commented:
You are reading it right. It's not that a small project requires two, it's just easier to generate. 

To limit it to a single reflection cache the code would need to tally everything that java might count against the constant pool in the generated class. It's a bit of a pain to do and probably a bit fragile on refactoring.
